src/main/java/net/clgd/ccemux/rendering/tror/TRoRRenderer.java

```diff
@@ -0,0 +1,213 @@
+package net.clgd.ccemux.rendering.tror;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.UnsupportedEncodingException;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingDeque;
+
+import net.clgd.ccemux.Utils;
+import net.clgd.ccemux.emulation.EmulatedComputer;
+import net.clgd.ccemux.emulation.EmulatedPalette;
+import net.clgd.ccemux.emulation.EmulatedTerminal;
+import net.clgd.ccemux.rendering.Renderer;
+import net.clgd.ccemux.rendering.RendererConfig;
+
+public class TRoRRenderer implements Renderer, EmulatedTerminal.Listener, EmulatedPalette.Listener {
+	private final EmulatedComputer computer;
+	private final RendererConfig config;
+
+	private final List<Listener> listeners = new ArrayList<>();
+
+	private final BlockingDeque<InputProvider.InputPacket> events;
+
+	private final Writer output;
+
+	public TRoRRenderer(EmulatedComputer computer, RendererConfig config) {
+		this.computer = computer;
+		this.config = config;
+
+		try {
+			output = new OutputStreamWriter(System.out, "UTF-8");
+		} catch (UnsupportedEncodingException e) {
+			throw new RuntimeException(e);
+		}
+
+		computer.terminal.addListener(this);
+		computer.terminal.getEmulatedPalette().addListener(this);
+
+		events = InputProvider.getStdinProvider().getQueue(computer);
+	}
+
+	@Override
+	public void addListener(Renderer.Listener listener) {
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeListener(Renderer.Listener listener) {
+		listeners.remove(listener);
+	}
+
+	@Override
+	public boolean isVisible() {
+		return true;
+	}
+
+	@Override
+	public void setVisible(boolean visible) {
+
+	}
+
+	@Override
+	public void dispose() {
+	}
+
+	@Override
+	public void onAdvance(double dt) {
+		InputProvider.InputPacket packet;
+		while ((packet = events.poll()) != null) {
+			switch (packet.code) {
+				case "EV": {
+					String payload = packet.payload;
+					int index = payload.indexOf(',');
+					if (index < 0) index = payload.length();
+
+					String event = payload.substring(0, index);
+					List<Object> args = new ArrayList<>();
+					index++;
+					while (index < payload.length()) {
+						char startChar = payload.charAt(index);
+						if (startChar == '"' || startChar == '\'') {
+							StringBuilder builder = new StringBuilder();
+							index++;
+							while (index < payload.length() - 1) {
+								char current = payload.charAt(index++);
+
+								if (current == startChar) {
+									break;
+								} else if (current == '\\') {
+									current = payload.charAt(index++);
+									if (current == 'a') builder.append((char) 0x7);
+									else if (current == 'b') builder.append('\b');
+									else if (current == 'f') builder.append('\f');
+									else if (current == 'n') builder.append('\n');
+									else if (current == 'r') builder.append('\r');
+									else if (current == 't') builder.append('\t');
```

Comment:
Of course it's ugly, it's parsing Lua :). I thought about using switch, but it ends up being 3x as many lines of code. I'm not fussed either way, just thought this was neater.